src/esperanza/walletextension.cpp

```diff
@@ -765,63 +767,27 @@ bool WalletExtension::SendSlash(const finalization::VoteRecord &vote1,
 
-  LOCK2(cs_main, m_enclosing_wallet.cs_wallet);
-  if (nIsValidatorEnabled) {
-    assert(validatorState);
-
-    LOCK(m_dependencies.GetFinalizationStateRepository().GetLock());
-    const CBlockIndex *tip_block_index = m_dependencies.GetActiveChain().GetTip();
-    const FinalizationState *fin_state = m_dependencies.GetFinalizationStateRepository().Find(*tip_block_index);
-    assert(fin_state);
-
-    const esperanza::Validator *validator = fin_state->GetValidator(validatorState->m_validator_address);
-    if (!validator) {
-      // this wallet doesn't have associated finalizer
-      // because no deposit from this wallet was made
-      return;
-    }
-
-    ValidatorStateWatchWriter validator_writer(*this);
+  if (!nIsValidatorEnabled) {
+    // finalizer is explicitly disabled
+    return;
+  }
 
-    if (fin_state->GetCurrentDynasty() < validator->m_start_dynasty) {
-      return;  // to early to vote
-    }
+  if (IsInitialBlockDownload()) {
+    // there is no reason to vote as such votes will be outdated
+    // and won't be included in the chain
+    return;
+  }
 
-    // TODO: UNIT-E: review if we need to keep track of the state
-    // as it seems checking against current dynasty is enough
-    if (fin_state->GetCurrentDynasty() <= validator->m_end_dynasty) {
-      if (validatorState->m_phase == +ValidatorState::Phase::WAITING_DEPOSIT_FINALIZATION) {
-        validatorState->m_phase = ValidatorState::Phase::IS_VALIDATING;
-        WriteValidatorStateToFile();
-        LogPrint(BCLog::FINALIZATION,
-                 "finalizer=%d phase changed to %s because start_dynasty=%d began\n",
-                 validator->m_validator_address.ToString(),
-                 validatorState->m_phase._to_string(),
-                 validator->m_start_dynasty);
-      }
-      assert(validatorState->m_phase == +ValidatorState::Phase::IS_VALIDATING);
-    } else {
-      if (validatorState->m_phase == +ValidatorState::Phase::IS_VALIDATING) {
-        validatorState->m_phase = ValidatorState::Phase::NOT_VALIDATING;
-        WriteValidatorStateToFile();
-        LogPrint(BCLog::FINALIZATION,
-                 "finalizer=%d phase changed to %s because end_dynasty=%d passed\n",
-                 validator->m_validator_address.ToString(),
-                 validatorState->m_phase._to_string(),
-                 validator->m_end_dynasty);
-      }
-      assert(validatorState->m_phase == +ValidatorState::Phase::NOT_VALIDATING);
-    }
+  LOCK2(cs_main, m_enclosing_wallet.cs_wallet);
+  assert(validatorState);
 
-    if (IsInitialBlockDownload()) {
-      // there is no reason to vote as such votes will be outdated
-      // and won't be included in the chain
-      return;
-    }
+  LOCK(m_dependencies.GetFinalizationStateRepository().GetLock());
+  const CBlockIndex *tip_block_index = m_dependencies.GetActiveChain().GetTip();
+  const FinalizationState *fin_state = m_dependencies.GetFinalizationStateRepository().Find(*tip_block_index);
```

Comment:
Can't this be just `GetTipState`?

Reply:
`VoteIfNeeded` needs to know the height. If we invoke `GetTipState` and `chainActive.Height` separately, there is a risk (but it's not possible now because of the locks) that `height` won't match with `GetTipState`.

Retrieving `CBlockIndex` and then getting all the meta information for it guarantees that meta information is exactly from that `CBlockIndex`.

Reply:
Ah, right. Missed `tip_block_index->nHeight` usage